### PR TITLE
feat(router): Add ability to set X-Request-Start header per app #42

### PR DESCRIPTION
**Pull request by Cryptophobia**
Closes https://github.com/teamhephy/router/issues/42

Still needs to be tested to make sure it works as intended for each application config.

Can you guys review @pdomagala and @felixbuenemann ?

### CODE REVIEW

**Comment by felixbuenemann** [comment]
@Cryptophobia I don't think it makes sense to have `appConfig.DisableRequestStartHeader`, it would be better to have `appConfig.RequestStartHeader` as a boolean.

You might want to have the `routerConfig.RequestStartHeader = false` and only enable `appConfig.RequestStartHeader = true` on some apps.

**Comment by pdomagala** [comment]
@Cryptophobia Exactly, I'm also worried that it'll interfere/override existing headers, so this solution (enable/disable per app) sounds legit.

File: nginx/config.go [hunk 1]
```diff
@@ -308,6 +308,9 @@ http {
 				proxy_set_header X-Request-Id $request_id;
 				proxy_set_header X-Correlation-Id $correlation_id;
 				{{ end }}
+				{{ if and $routerConfig.RequestStartHeader (ne $appConfig.DisableRequestStartHeader true) }}
+				proxy_set_header X-Request-Start "t=${msec}";
+				{{end}}
```
Comment (kingdonb) [inline nginx/config.go]: This line misses some spaces to conform with the surrounding style

**Comment by Cryptophobia** [comment]
Thank you for reviewing this @pdomagala and @felixbuenemann .

So let's go with what @felixbuenemann mentioned above:

> So if appConfig.RequestStartHeader is nil, we use the value from routerConfig.RequestStartHeader, but if it is either true or false it overrides the routerConfig value.

File: nginx/config.go (hunk 1 above)
Comment (Cryptophobia) [reply nginx/config.go]: Fixed. Thanks! :+1:

File: nginx/config.go [hunk 2]
```diff
@@ -308,6 +308,9 @@ http {
 				proxy_set_header X-Request-Id $request_id;
 				proxy_set_header X-Correlation-Id $correlation_id;
 				{{ end }}
+				{{ if and $routerConfig.RequestStartHeader (or $appConfig.RequestStartHeader (eq $appConfig.RequestStartHeader nil))}}
```
Comment (Cryptophobia) [inline nginx/config.go]: @pdomagala @felixbuenemann Can you guys take a look at this? Is this the logic we are looking for?

File: README.md [hunk 3]
```diff
@@ -283,6 +284,7 @@ _Note that Kubernetes annotation maps are all of Go type `map[string]string`.  A
 | <a name="app-connect-timeout"></a>routable application | service | [router.deis.io/connectTimeout](#app-connect-timeout) | `"30s"` | nginx `proxy_connect_timeout` setting expressed in units `ms`, `s`, `m`, `h`, `d`, `w`, `M`, or `y`. |
 | <a name="app-tcp-timeout"></a>routable application | service | [router.deis.io/tcpTimeout](#app-tcp-timeout) | router's `defaultTimeout` | nginx `proxy_send_timeout` and `proxy_read_timeout` settings expressed in units `ms`, `s`, `m`, `h`, `d`, `w`, `M`, or `y`. |
 | <a name="app-maintenance"></a>routable application | service | [router.deis.io/maintenance](#app-maintenance) | `"false"` | Whether the app is under maintenance so that all traffic for this app is redirected to a static maintenance page with an error code of `503`. |
+| <a name="app-request-start-header"></a>routable application | service | [router.deis.io/requestStartHeader](#app-request-start-header) | N/A | Whether to disable adding the `X-Request-Start` headers to the application when X-Request-Header is set globally. Must be set to `false` in order to opt out of the global behavior for all apps. |
```
Comment (Cryptophobia) [inline README.md]: Does this also make sense in terms of the description?

**Comment by Cryptophobia** [comment]
I have tested this and it works perfectly the way we want. I have kept he old naming because it makes more sense if the annotation is called `disableRequestStartHeader` since the values of it can be either `true` or `false`. If the header is missing from the appConfig then we just use the global value override.

**Comment by felixbuenemann** [comment]
@Cryptophobia The thing I don't like about this implementation, is that if you want to enable the header for a single app, you would need to set `routerConfig.requestStartHeader = true` and then set `appConfig.disableRequestStartHeader = false` for every app, except the one app, where we want the header.

With the logic I proposed you would have just set `appConfig.requestStratHeader = true` for the one app, where you want it enabled.

I don't think it's a dealbreaker, since it most cases you would be fine with having the header active for all apps, even if they don't need it.

**Comment by Cryptophobia** [comment]
> I don't think it's a dealbreaker, since it most cases you would be fine with having the header active for all apps, even if they don't need it.

This is what I thought from the beginning. Having this extra header even if not used by anything (like Datadog, etc.), is it actually a burden?

**Comment by felixbuenemann** [comment]
> > I don't think it's a dealbreaker, since it most cases you would be fine with having the header active for all apps, even if they don't need it.
> 
> This is what I thought from the beginning. Having this extra header even if not used by anything (like Datadog, etc.), is it actually a burden?

Nope, let's ship it like this!

**Pull request merged**